Revit_Core_Engine/Query/OpeningSurface.cs

```diff
@@ -54,12 +63,102 @@ public static ISurface OpeningSurface(this FamilyInstance familyInstance, HostOb
             {
                 BoundingBoxIntersectsFilter bbif = new BoundingBoxIntersectsFilter(new Outline(bbox.Min, bbox.Max));
                 hosts = new FilteredElementCollector(familyInstance.Document).OfClass(typeof(HostObject)).WherePasses(bbif).Cast<HostObject>().ToList();
-                hosts = hosts.Where(x => x.FindInserts(true, true, true, true).Any(y => y.IntegerValue == familyInstance.Id.IntegerValue)).ToList();
+                hosts = hosts.Where(x => x.FindInserts(true, true, true, true).Any(y => (y.IntegerValue == familyInstance.Id.IntegerValue)|| (y.IntegerValue == parentId))).ToList();
             }
 
             List<ISurface> surfaces = new List<ISurface>();
 
-            if (hosts.Count == 0)
+            if (parentElem != null) // Indicates element is a nested instance and needs different approach to extract location
+            {
+                Document doc = familyInstance.Document;
+                Transaction t = new Transaction(doc);
+
+                FailureHandlingOptions failureHandlingOptions = t.GetFailureHandlingOptions().SetClearAfterRollback(true);
+                t.Start("Temp Delete Inserts");
+                List<ElementId> inserts = hosts.SelectMany(x => x.FindInserts(true, true, true, true)).Distinct().Where(x => x.IntegerValue != -1).ToList();
+                ParameterSet instanceParams = familyInstance.Parameters;
+                List<RevitParameter> bhomParams = new List<RevitParameter>();
+                foreach (Parameter parameter in instanceParams)
+                {
+                    RevitParameter bhomParam = Convert.ParameterFromRevit(parameter);
+                    bhomParams.Add(bhomParam);
+                }
```

Comment:
I do not understand this part of the code - why would you convert the Revit parameters to BHoM parameter objects and then do matching with hard coded names (_Sill Height_, _Head Height_)? This does not look sustainable and I would argue whether it is necessary.

Reply:
I needed a way to effectively copy parameters, and as far as I could tell this is the best available method so that we can take advantage of the SetParameters method to manage setting of different parameter types. The hardcoded exclusion of Sill Height and Head Height is due to these causing errors if set through the api; they do not need setting anyhow as setting location and the other params places the instance in the exact location without touching these two parameters. If we have a better means of effectively copying revit params from one element and assigning to another I'm all ears!

Reply:
I believe you could simply use in-built Revit get/set parameter functionality - no need to convert to BHoM parameters, it is just an overhead as far as I understand. If you need to skip any parameters, I recommend using [BuiltInParameter enum](https://www.revitapidocs.com/2020/fb011c91-be7e-f737-28c7-3f1e1917a0e0.htm) instead of string.

Reply:
I've removed the redundant to and from bhom parameter portion. This required boilerplating the switches for the different Revit parameter value types in order to get the param values to then set them. Perhaps worth creating a method that handles that for us? Revit API of course does not provide for something as simple as SetParameter(element, parameter) which would streamline the code here.

Reply:
BuiltInEnum is now used for the Head and Sill Height exclusion as well, good call there!